### DISCUSSION TITLE
Deploy contract using safe wallet

### DISCUSSION THREAD

**Discussion by swatijindal12** [safe-global/safe-apps-sdk#523] [General]
I created a safe wallet on goerli testnet. Can we deploy new smart contract using safe wallet?

**Comment by DaniSomoza** [comment]
Hi @swatijindal12 

Yes, while your Safe Account itself is not an Externally Owned Account (EOA) and therefore cannot natively deploy a smart contract directly like an EOA would, it's still possible to deploy new contracts using your Safe.

To accomplish this, you would need to interact with a contract that is designed to create other contracts. We have a [`CreateCall` contract](https://github.com/safe-global/safe-contracts/blob/main/contracts/libraries/CreateCall.sol) that has two functions named `performCreate` & `performCreate2` that uses the `CREATE` or `CREATE2` opcodes to deploy new contracts.

To interact with the `CreateCall` contract (or any other contract) from your Safe Account, you can our the **Transaction Builder Safe App**. It can be found in our Apps Section:

![Captura de pantalla 2023-10-03 a las 18 34 04](https://github.com/safe-global/safe-apps-sdk/assets/26763673/626e3668-acb6-479f-8017-d64d7b4bd884)

Within the Transaction Builder, you can input both the contract details and the specific function parameters you aim to call. This enables you to invoke the performCreate function on the CreateCall contract:

On the Goerli network, the address for the CreateCall contract is [`0x7cbB62EaA69F79e6873cD1ecB2392971036cFAa4`](https://goerli.etherscan.io/address/0x7cbB62EaA69F79e6873cD1ecB2392971036cFAa4#code). For the list of addresses across all chains, you can refer to our [safe-deployments repository](https://github.com/safe-global/safe-deployments/blob/main/src/assets/v1.3.0/create_call.json).

First, , input the CreateCall contract address:
![Captura de pantalla 2023-10-03 a las 18 42 49](https://github.com/safe-global/safe-apps-sdk/assets/26763673/2999cf76-68c5-4b74-b88d-0a996d8f0dc9)

Then, select the desired method for creating your contract (`performCreate` or `performCreate2`) and provide the value and the bytecode of the contract that you want to deploy:

![Captura de pantalla 2023-10-03 a las 18 43 57](https://github.com/safe-global/safe-apps-sdk/assets/26763673/c3cccd97-0a84-4ec3-a61a-e3d2b5704428)

Now you can click on `Add transaction` button and Create & execute the batch. 

For reference, you can view this [transaction with this example here](https://goerli.etherscan.io/tx/0x56a8f107ec8f28c8968c150fe53717d64ce6051ee887e648fdaafb62c6c9fe3d).

I deployed [this smart contract](https://goerli.etherscan.io/address/0x4e81f1f16a3CE3F0616A772b906C50486d7Ed2bD) using my [Safe Account](https://goerli.etherscan.io/address/0x9Fe2725B42a17395f257A0fc0Bf3Ddb703B4c786) in [this transaction](https://goerli.etherscan.io/tx/0x56a8f107ec8f28c8968c150fe53717d64ce6051ee887e648fdaafb62c6c9fe3d)


Note: if you check the owner of the test contract that I deployed using my Safe Account you can see that the owner is the `CreateCall` contract instead of the Safe Account contract:

![Captura de pantalla 2023-10-03 a las 18 54 23](https://github.com/safe-global/safe-apps-sdk/assets/26763673/4050c821-e59c-44e3-a9e7-2d8f7b090516)


This arises because I executed a **call** from my Safe Account to the `CreateCall` contract rather than a **delegateCall**. If this mirrors your use case, please let me know.

Thanks!

**Comment by DaniSomoza** [comment]
If you need to deploy via delegateCall, you can utilize our protocol-kit:

```javascript
import Safe, { EthersAdapter, getCreateCallContract } from '@safe-global/protocol-kit'
import { OperationType } from '@safe-global/safe-core-sdk-types'
import { ethers } from 'ethers'

// Goerli config
const config = {
  RPC_URL: 'https://goerli.infura.io/v3/9aa3d95b3bc440fa88ea12eaa4456161',
  SAFE_SIGNER_PRIVATE_KEY: '<SIGNER_PK>',
  SAFE_ADDRESS: '<YOUR_SAFE_ADDRESS>' // Goerli Safe 1/1
}

const contractBytecode =
  '0x6080604052348015600f57600080fd5b50600080546001600160a01b0319163317905560a0806100306000396000f3fe6080604052348015600f57600080fd5b506004361060285760003560e01c8063893d20e814602d575b600080fd5b60336047565b604051603e91906056565b60405180910390f35b6000546001600160a01b031690565b6001600160a01b039190911681526020019056fea2646970667358221220898723a0669d8eb946185c275f6547a97576c55c9bcee1bae7035f9d6bd2f44364736f6c63430008000033'

// Safe version
const safeVersion = '1.3.0'

async function main() {
  console.log(`Deployment of a contract from a Safe Account:`)

  // SDK Initialization
  const provider = new ethers.providers.JsonRpcProvider(config.RPC_URL)
  const signer = new ethers.Wallet(config.SAFE_SIGNER_PRIVATE_KEY, provider)

  const ethAdapter = new EthersAdapter({
    ethers,
    signerOrProvider: signer
  })

  const safeSDK: Safe = await Safe.create({
    ethAdapter,
    safeAddress: config.SAFE_ADDRESS
  })

  // Here we add the deployment of the contract with my Safe Account

  const createCallContract = await getCreateCallContract({ ethAdapter, safeVersion })
  const createCallContractAddress = await createCallContract.getAddress()
  const value = '0'
  const data = createCallContract.encode('performCreate', [value, contractBytecode])

  // transaction with the deployment (it is a delegateCall from the Safe Account to the performCreate method of the createCall contract)
  const transaction = {
    to: createCallContractAddress,
    value,
    data,
    operation: OperationType.DelegateCall
  }

  const safeTransaction = await safeSDK.createTransaction({ safeTransactionData: transaction })

  // we execute the deployment contract transaction
  const { transactionResponse } = await safeSDK.executeTransaction(safeTransaction)

  const receipt = await transactionResponse?.wait()

  console.log('Transaction executed:')
  console.log(`https://goerli.etherscan.io/tx/${receipt.transactionHash}`)
}

main()

```

You can check [the transaction details here](https://goerli.etherscan.io/tx/0x4a8c65d8c7c743994841c5b9ee021fd151eceb752cd7dab9128c186b8179845a)

You can see that when deploying via **delegateCall**, the owner of the contract becomes the Safe Contract itself, rather than the `CreateCall` contract:

![Captura de pantalla 2023-10-03 a las 19 17 33](https://github.com/safe-global/safe-apps-sdk/assets/26763673/b184c053-c719-453c-a38f-fca1d4f222f2)

Let me know if you need more help or have any other questions!

**Reply by swatijindal12** [reply]
If Number of owners is 3 & threshold value is 2 then how we deploy a contract?

**Reply by DaniSomoza** [reply]
In this case you need to sign the transaction with 2 owners using `signTransaction` :

```javascript
const safeTransaction = await safeSDK.createTransaction({ safeTransactionData: transaction })

// add first signature
let signedSafeTransaction = await safeSDK.signTransaction(safeTransaction)

// connect the other owner
const ethAdapter = new EthersAdapter({
  ethers,
  signerOrProvider: new ethers.Wallet(config.SAFE_SIGNER_2_PRIVATE_KEY, provider)
})

const safeWithNewOwner = await safeSDK.connect({
  ethAdapter
})

// add second signature
signedSafeTransaction = await safeWithNewOwner.signTransaction(signedSafeTransaction)

// we execute the deployment contract transaction
const { transactionResponse } = await safeWithNewOwner.executeTransaction(signedSafeTransaction)

```

**Reply by swatijindal12** [reply]
Okay. Thanks i will try

**Reply by swatijindal12** [reply]
I tried using this but still got error "there is 1 signature missing"

const {
  EthersAdapter,
  getCreateCallContract,
} = require('@safe-global/protocol-kit')
const Safe = require('@safe-global/protocol-kit').default
const { OperationType } = require('@safe-global/safe-core-sdk-types')
const { ethers } = require('ethers')

// Goerli config
const config = {
  RPC_URL:
    'https://eth-goerli.g.alchemy.com/v2/....',
  SAFE_SIGNER_PRIVATE_KEY:
    '0x....',
  SAFE_SIGNER_2_PRIVATE_KEY:
    '0x....',
  SAFE_ADDRESS: '0x0a8Bb367aDa8a7A3Bd3e3EE198f4d6552a1caE41', // Goerli Safe 1/1
}

const contractBytecode =
  '0x608060405234801561000f575f80fd5b506040516106ed3803806106ed833981810160405281019061003191906101f4565b805f836040516100419190610292565b90815260200160405180910390208190555050506102a8565b5f604051905090565b5f80fd5b5f80fd5b5f80fd5b5f80fd5b5f601f19601f8301169050919050565b7f4e487b71000000000000000000000000000000000000000000000000000000005f52604160045260245ffd5b6100b982610073565b810181811067ffffffffffffffff821117156100d8576100d7610083565b5b80604052505050565b5f6100ea61005a565b90506100f682826100b0565b919050565b5f67ffffffffffffffff82111561011557610114610083565b5b61011e82610073565b9050602081019050919050565b5f5b8381101561014857808201518184015260208101905061012d565b5f8484015250505050565b5f610165610160846100fb565b6100e1565b9050828152602081018484840111156101815761018061006f565b5b61018c84828561012b565b509392505050565b5f82601f8301126101a8576101a761006b565b5b81516101b8848260208601610153565b91505092915050565b5f819050919050565b6101d3816101c1565b81146101dd575f80fd5b50565b5f815190506101ee816101ca565b92915050565b5f806040838503121561020a57610209610063565b5b5f83015167ffffffffffffffff81111561022757610226610067565b5b61023385828601610194565b9250506020610244858286016101e0565b9150509250929050565b5f81519050919050565b5f81905092915050565b5f61026c8261024e565b6102768185610258565b935061028681856020860161012b565b80840191505092915050565b5f61029d8284610262565b915081905092915050565b610438806102b55f395ff3fe608060405234801561000f575f80fd5b506004361061003f575f3560e01c806314b3ee68146100435780639f11592114610073578063fd8111e91461008f575b5f80fd5b61005d60048036038101906100589190610284565b6100bf565b60405161006a91906102e3565b60405180910390f35b61008d60048036038101906100889190610326565b6100e5565b005b6100a960048036038101906100a49190610284565b61010b565b6040516100b691906102e3565b60405180910390f35b5f80826040516100cf91906103ec565b9081526020016040518091039020549050919050565b805f836040516100f591906103ec565b9081526020016040518091039020819055505050565b5f818051602081018201805184825260208301602085012081835280955050505050505f915090505481565b5f604051905090565b5f80fd5b5f80fd5b5f80fd5b5f80fd5b5f601f19601f8301169050919050565b7f4e487b71000000000000000000000000000000000000000000000000000000005f52604160045260245ffd5b61019682610150565b810181811067ffffffffffffffff821117156101b5576101b4610160565b5b80604052505050565b5f6101c7610137565b90506101d3828261018d565b919050565b5f67ffffffffffffffff8211156101f2576101f1610160565b5b6101fb82610150565b9050602081019050919050565b828183375f83830152505050565b5f610228610223846101d8565b6101be565b9050828152602081018484840111156102445761024361014c565b5b61024f848285610208565b509392505050565b5f82601f83011261026b5761026a610148565b5b813561027b848260208601610216565b91505092915050565b5f6020828403121561029957610298610140565b5b5f82013567ffffffffffffffff8111156102b6576102b5610144565b5b6102c284828501610257565b91505092915050565b5f819050919050565b6102dd816102cb565b82525050565b5f6020820190506102f65f8301846102d4565b92915050565b610305816102cb565b811461030f575f80fd5b50565b5f81359050610320816102fc565b92915050565b5f806040838503121561033c5761033b610140565b5b5f83013567ffffffffffffffff81111561035957610358610144565b5b61036585828601610257565b925050602061037685828601610312565b9150509250929050565b5f81519050919050565b5f81905092915050565b5f5b838110156103b1578082015181840152602081019050610396565b5f8484015250505050565b5f6103c682610380565b6103d0818561038a565b93506103e0818560208601610394565b80840191505092915050565b5f6103f782846103bc565b91508190509291505056fea2646970667358221220c01473f66ae07e24c0254030668e79bf101a07b7433c4349521974f7f788d56164736f6c63430008150033'
// Safe version
const safeVersion = '1.3.0'

async function deployContract() {
  console.log(`Deployment of a contract from a Safe Account:`)

  // SDK Initialization
  const provider = new ethers.providers.JsonRpcProvider(config.RPC_URL)
  const signer = new ethers.Wallet(config.SAFE_SIGNER_PRIVATE_KEY, provider)

  const ethAdapter = new EthersAdapter({
    ethers,
    signerOrProvider: signer,
  })

  const safeSDK = await Safe.create({
    ethAdapter,
    safeAddress: config.SAFE_ADDRESS,
  })

  // Here we add the deployment of the contract with my Safe Account

  const createCallContract = await getCreateCallContract({
    ethAdapter,
    safeVersion,
  })
  const createCallContractAddress = await createCallContract.getAddress()
  const value = '0'
  const data = createCallContract.encode('performCreate', [
    value,
    contractBytecode,
  ])

  // transaction with the deployment (it is a delegateCall from the Safe Account to the performCreate method of the createCall contract)
  const transaction = {
    to: createCallContractAddress,
    value,
    data,
    operation: OperationType.DelegateCall,
  }

  const safeTransaction = await safeSDK.createTransaction({
    safeTransactionData: transaction,
  })

  // add first signature
  let signedSafeTransaction = await safeSDK.signTransaction(safeTransaction)

  // connect the other owner
  const ethAdapter2 = new EthersAdapter({
    ethers,
    signerOrProvider: new ethers.Wallet(
      config.SAFE_SIGNER_2_PRIVATE_KEY,
      provider,
    ),
  })

  const safeWithNewOwner = await safeSDK.connect({
    ethAdapter2,
  })

  // add second signature
  signedSafeTransaction = await safeWithNewOwner.signTransaction(
    signedSafeTransaction,
  )

  // we execute the deployment contract transaction
  const { transactionResponse } = await safeWithNewOwner.executeTransaction(
    signedSafeTransaction,
  )

  const receipt = await transactionResponse?.wait()

  console.log('Transaction executed:')
  console.log(`https://goerli.etherscan.io/tx/${receipt.transactionHash}`)
}

deployContract()

**Comment by swatijindal12** [comment]
But when i deploy a contract using transaction builder i got error "Failed. Execution reverted: GS013."  In performCreate i passed value as a zero & passed bytecode of the contract. 
![Screenshot from 2023-10-03 22-52-02](https://github.com/safe-global/safe-apps-sdk/assets/55221427/735d28cf-bfc6-4f0d-94d4-4ae29591546f)

**Reply by DaniSomoza** [reply]
Could you provide me more details about the Safe and the deploymentData that you are using?

**Reply by swatijindal12** [reply]
yes sure. 
Deployment data: 608060405234801562000010575f80fd5b5060405162001dc738038062001dc783398181016040528101906200003691906200038b565b6040518060400160405280601481526020017f537461724475636b7320436170752d546f6b656e0000000000000000000000008152506040518060400160405280600381526020017f53435400000000000000000000000000000000000000000000000000000000008152508160039081620000b3919062000616565b508060049081620000c5919062000616565b505050620000e8620000dc6200011660201b60201c565b6200011d60201b60201c565b6200010f33670de0b6b3a76400008362000103919062000727565b620001e060201b60201c565b5062000855565b5f33905090565b5f60055f9054906101000a900473ffffffffffffffffffffffffffffffffffffffff1690508160055f6101000a81548173ffffffffffffffffffffffffffffffffffffffff021916908373ffffffffffffffffffffffffffffffffffffffff1602179055508173ffffffffffffffffffffffffffffffffffffffff168173ffffffffffffffffffffffffffffffffffffffff167f8be0079c531659141344cd1fd0a4f28419497f9722a3daafe3b4186f6b6457e060405160405180910390a35050565b5f73ffffffffffffffffffffffffffffffffffffffff168273ffffffffffffffffffffffffffffffffffffffff160362000251576040517f08c379a00000000000000000000000000000000000000000000000000000000081526004016200024890620007cf565b60405180910390fd5b620002645f83836200034560201b60201c565b8060025f828254620002779190620007ef565b92505081905550805f808473ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff1681526020019081526020015f205f82825401925050819055508173ffffffffffffffffffffffffffffffffffffffff165f73ffffffffffffffffffffffffffffffffffffffff167fddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef836040516200032691906200083a565b60405180910390a3620003415f83836200034a60201b60201c565b5050565b505050565b505050565b5f80fd5b5f819050919050565b620003678162000353565b811462000372575f80fd5b50565b5f8151905062000385816200035c565b92915050565b5f60208284031215620003a357620003a26200034f565b5b5f620003b28482850162000375565b91505092915050565b5f81519050919050565b7f4e487b71000000000000000000000000000000000000000000000000000000005f52604160045260245ffd5b7f4e487b71000000000000000000000000000000000000000000000000000000005f52602260045260245ffd5b5f60028204905060018216806200043757607f821691505b6020821081036200044d576200044c620003f2565b5b50919050565b5f819050815f5260205f209050919050565b5f6020601f8301049050919050565b5f82821b905092915050565b5f60088302620004b17fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff8262000474565b620004bd868362000474565b95508019841693508086168417925050509392505050565b5f819050919050565b5f620004fe620004f8620004f28462000353565b620004d5565b62000353565b9050919050565b5f819050919050565b6200051983620004de565b62000531620005288262000505565b84845462000480565b825550505050565b5f90565b6200054762000539565b620005548184846200050e565b505050565b5b818110156200057b576200056f5f826200053d565b6001810190506200055a565b5050565b601f821115620005ca57620005948162000453565b6200059f8462000465565b81016020851015620005af578190505b620005c7620005be8562000465565b83018262000559565b50505b505050565b5f82821c905092915050565b5f620005ec5f1984600802620005cf565b1980831691505092915050565b5f620006068383620005db565b9150826002028217905092915050565b6200062182620003bb565b67ffffffffffffffff8111156200063d576200063c620003c5565b5b6200064982546200041f565b620006568282856200057f565b5f60209050601f8311600181146200068c575f841562000677578287015190505b620006838582620005f9565b865550620006f2565b601f1984166200069c8662000453565b5f5b82811015620006c5578489015182556001820191506020850194506020810190506200069e565b86831015620006e55784890151620006e1601f891682620005db565b8355505b6001600288020188555050505b505050505050565b7f4e487b71000000000000000000000000000000000000000000000000000000005f52601160045260245ffd5b5f620007338262000353565b9150620007408362000353565b9250828202620007508162000353565b915082820484148315176200076a5762000769620006fa565b5b5092915050565b5f82825260208201905092915050565b7f45524332303a206d696e7420746f20746865207a65726f2061646472657373005f82015250565b5f620007b7601f8362000771565b9150620007c48262000781565b602082019050919050565b5f6020820190508181035f830152620007e881620007a9565b9050919050565b5f620007fb8262000353565b9150620008088362000353565b9250828201905080821115620008235762000822620006fa565b5b92915050565b620008348162000353565b82525050565b5f6020820190506200084f5f83018462000829565b92915050565b61156480620008635f395ff3fe608060405234801561000f575f80fd5b50600436106100e8575f3560e01c8063715018a61161008a578063a457c2d711610064578063a457c2d71461024c578063a9059cbb1461027c578063dd62ed3e146102ac578063f2fde38b146102dc576100e8565b8063715018a6146102065780638da5cb5b1461021057806395d89b411461022e576100e8565b806323b872dd116100c657806323b872dd14610158578063313ce5671461018857806339509351146101a657806370a08231146101d6576100e8565b806306fdde03146100ec578063095ea7b31461010a57806318160ddd1461013a575b5f80fd5b6100f46102f8565b6040516101019190610d60565b60405180910390f35b610124600480360381019061011f9190610e11565b610388565b6040516101319190610e69565b60405180910390f35b6101426103aa565b60405161014f9190610e91565b60405180910390f35b610172600480360381019061016d9190610eaa565b6103b3565b60405161017f9190610e69565b60405180910390f35b6101906103e1565b60405161019d9190610f15565b60405180910390f35b6101c060048036038101906101bb9190610e11565b6103e9565b6040516101cd9190610e69565b60405180910390f35b6101f060048036038101906101eb9190610f2e565b61041f565b6040516101fd9190610e91565b60405180910390f35b61020e610464565b005b610218610477565b6040516102259190610f68565b60405180910390f35b61023661049f565b6040516102439190610d60565b60405180910390f35b61026660048036038101906102619190610e11565b61052f565b6040516102739190610e69565b60405180910390f35b61029660048036038101906102919190610e11565b6105a4565b6040516102a39190610e69565b60405180910390f35b6102c660048036038101906102c19190610f81565b6105c6565b6040516102d39190610e91565b60405180910390f35b6102f660048036038101906102f19190610f2e565b610648565b005b60606003805461030790610fec565b80601f016020809104026020016040519081016040528092919081815260200182805461033390610fec565b801561037e5780601f106103555761010080835404028352916020019161037e565b820191905f5260205f20905b81548152906001019060200180831161036157829003601f168201915b5050505050905090565b5f806103926106ca565b905061039f8185856106d1565b600191505092915050565b5f600254905090565b5f806103bd6106ca565b90506103ca858285610894565b6103d585858561091f565b60019150509392505050565b5f6012905090565b5f806103f36106ca565b905061041481858561040585896105c6565b61040f9190611049565b6106d1565b600191505092915050565b5f805f8373ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff1681526020019081526020015f20549050919050565b61046c610b8b565b6104755f610c09565b565b5f60055f9054906101000a900473ffffffffffffffffffffffffffffffffffffffff16905090565b6060600480546104ae90610fec565b80601f01602080910402602001604051908101604052809291908181526020018280546104da90610fec565b80156105255780601f106104fc57610100808354040283529160200191610525565b820191905f5260205f20905b81548152906001019060200180831161050857829003601f168201915b5050505050905090565b5f806105396106ca565b90505f61054682866105c6565b90508381101561058b576040517f08c379a0000000000000000000000000000000000000000000000000000000008152600401610582906110ec565b60405180910390fd5b61059882868684036106d1565b60019250505092915050565b5f806105ae6106ca565b90506105bb81858561091f565b600191505092915050565b5f60015f8473ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff1681526020019081526020015f205f8373ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff1681526020019081526020015f2054905092915050565b610650610b8b565b5f73ffffffffffffffffffffffffffffffffffffffff168173ffffffffffffffffffffffffffffffffffffffff16036106be576040517f08c379a00000000000000000000000000000000000000000000000000000000081526004016106b59061117a565b60405180910390fd5b6106c781610c09565b50565b5f33905090565b5f73ffffffffffffffffffffffffffffffffffffffff168373ffffffffffffffffffffffffffffffffffffffff160361073f576040517f08c379a000000000000000000000000000000000000000000000000000000000815260040161073690611208565b60405180910390fd5b5f73ffffffffffffffffffffffffffffffffffffffff168273ffffffffffffffffffffffffffffffffffffffff16036107ad576040517f08c379a00000000000000000000000000000000000000000000000000000000081526004016107a490611296565b60405180910390fd5b8060015f8573ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff1681526020019081526020015f205f8473ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff1681526020019081526020015f20819055508173ffffffffffffffffffffffffffffffffffffffff168373ffffffffffffffffffffffffffffffffffffffff167f8c5be1e5ebec7d5bd14f71427d1e84f3dd0314c0f7b2291e5b200ac8c7c3b925836040516108879190610e91565b60405180910390a3505050565b5f61089f84846105c6565b90507fffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff8114610919578181101561090b576040517f08c379a0000000000000000000000000000000000000000000000000000000008152600401610902906112fe565b60405180910390fd5b61091884848484036106d1565b5b50505050565b5f73ffffffffffffffffffffffffffffffffffffffff168373ffffffffffffffffffffffffffffffffffffffff160361098d576040517f08c379a00000000000000000000000000000000000000000000000000000000081526004016109849061138c565b60405180910390fd5b5f73ffffffffffffffffffffffffffffffffffffffff168273ffffffffffffffffffffffffffffffffffffffff16036109fb576040517f08c379a00000000000000000000000000000000000000000000000000000000081526004016109f29061141a565b60405180910390fd5b610a06838383610ccc565b5f805f8573ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff1681526020019081526020015f2054905081811015610a89576040517f08c379a0000000000000000000000000000000000000000000000000000000008152600401610a80906114a8565b60405180910390fd5b8181035f808673ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff1681526020019081526020015f2081905550815f808573ffffffffffffffffffffffffffffffffffffffff1673ffffffffffffffffffffffffffffffffffffffff1681526020019081526020015f205f82825401925050819055508273ffffffffffffffffffffffffffffffffffffffff168473ffffffffffffffffffffffffffffffffffffffff167fddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef84604051610b729190610e91565b60405180910390a3610b85848484610cd1565b50505050565b610b936106ca565b73ffffffffffffffffffffffffffffffffffffffff16610bb1610477565b73ffffffffffffffffffffffffffffffffffffffff1614610c07576040517f08c379a0000000000000000000000000000000000000000000000000000000008152600401610bfe90611510565b60405180910390fd5b565b5f60055f9054906101000a900473ffffffffffffffffffffffffffffffffffffffff1690508160055f6101000a81548173ffffffffffffffffffffffffffffffffffffffff021916908373ffffffffffffffffffffffffffffffffffffffff1602179055508173ffffffffffffffffffffffffffffffffffffffff168173ffffffffffffffffffffffffffffffffffffffff167f8be0079c531659141344cd1fd0a4f28419497f9722a3daafe3b4186f6b6457e060405160405180910390a35050565b505050565b505050565b5f81519050919050565b5f82825260208201905092915050565b5f5b83811015610d0d578082015181840152602081019050610cf2565b5f8484015250505050565b5f601f19601f8301169050919050565b5f610d3282610cd6565b610d3c8185610ce0565b9350610d4c818560208601610cf0565b610d5581610d18565b840191505092915050565b5f6020820190508181035f830152610d788184610d28565b905092915050565b5f80fd5b5f73ffffffffffffffffffffffffffffffffffffffff82169050919050565b5f610dad82610d84565b9050919050565b610dbd81610da3565b8114610dc7575f80fd5b50565b5f81359050610dd881610db4565b92915050565b5f819050919050565b610df081610dde565b8114610dfa575f80fd5b50565b5f81359050610e0b81610de7565b92915050565b5f8060408385031215610e2757610e26610d80565b5b5f610e3485828601610dca565b9250506020610e4585828601610dfd565b9150509250929050565b5f8115159050919050565b610e6381610e4f565b82525050565b5f602082019050610e7c5f830184610e5a565b92915050565b610e8b81610dde565b82525050565b5f602082019050610ea45f830184610e82565b92915050565b5f805f60608486031215610ec157610ec0610d80565b5b5f610ece86828701610dca565b9350506020610edf86828701610dca565b9250506040610ef086828701610dfd565b9150509250925092565b5f60ff82169050919050565b610f0f81610efa565b82525050565b5f602082019050610f285f830184610f06565b92915050565b5f60208284031215610f4357610f42610d80565b5b5f610f5084828501610dca565b91505092915050565b610f6281610da3565b82525050565b5f602082019050610f7b5f830184610f59565b92915050565b5f8060408385031215610f9757610f96610d80565b5b5f610fa485828601610dca565b9250506020610fb585828601610dca565b9150509250929050565b7f4e487b71000000000000000000000000000000000000000000000000000000005f52602260045260245ffd5b5f600282049050600182168061100357607f821691505b60208210810361101657611015610fbf565b5b50919050565b7f4e487b71000000000000000000000000000000000000000000000000000000005f52601160045260245ffd5b5f61105382610dde565b915061105e83610dde565b92508282019050808211156110765761107561101c565b5b92915050565b7f45524332303a2064656372656173656420616c6c6f77616e63652062656c6f775f8201527f207a65726f000000000000000000000000000000000000000000000000000000602082015250565b5f6110d6602583610ce0565b91506110e18261107c565b604082019050919050565b5f6020820190508181035f830152611103816110ca565b9050919050565b7f4f776e61626c653a206e6577206f776e657220697320746865207a65726f20615f8201527f6464726573730000000000000000000000000000000000000000000000000000602082015250565b5f611164602683610ce0565b915061116f8261110a565b604082019050919050565b5f6020820190508181035f83015261119181611158565b9050919050565b7f45524332303a20617070726f76652066726f6d20746865207a65726f206164645f8201527f7265737300000000000000000000000000000000000000000000000000000000602082015250565b5f6111f2602483610ce0565b91506111fd82611198565b604082019050919050565b5f6020820190508181035f83015261121f816111e6565b9050919050565b7f45524332303a20617070726f766520746f20746865207a65726f2061646472655f8201527f7373000000000000000000000000000000000000000000000000000000000000602082015250565b5f611280602283610ce0565b915061128b82611226565b604082019050919050565b5f6020820190508181035f8301526112ad81611274565b9050919050565b7f45524332303a20696e73756666696369656e7420616c6c6f77616e63650000005f82015250565b5f6112e8601d83610ce0565b91506112f3826112b4565b602082019050919050565b5f6020820190508181035f830152611315816112dc565b9050919050565b7f45524332303a207472616e736665722066726f6d20746865207a65726f2061645f8201527f6472657373000000000000000000000000000000000000000000000000000000602082015250565b5f611376602583610ce0565b91506113818261131c565b604082019050919050565b5f6020820190508181035f8301526113a38161136a565b9050919050565b7f45524332303a207472616e7366657220746f20746865207a65726f20616464725f8201527f6573730000000000000000000000000000000000000000000000000000000000602082015250565b5f611404602383610ce0565b915061140f826113aa565b604082019050919050565b5f6020820190508181035f830152611431816113f8565b9050919050565b7f45524332303a207472616e7366657220616d6f756e74206578636565647320625f8201527f616c616e63650000000000000000000000000000000000000000000000000000602082015250565b5f611492602683610ce0565b915061149d82611438565b604082019050919050565b5f6020820190508181035f8301526114bf81611486565b9050919050565b7f4f776e61626c653a2063616c6c6572206973206e6f7420746865206f776e65725f82015250565b5f6114fa602083610ce0565b9150611505826114c6565b602082019050919050565b5f6020820190508181035f830152611527816114ee565b905091905056fea2646970667358221220faf09d0cabef0248643b46c1e9af7ebca7e9cc98f1aa61e69ee9080e6e24c15e64736f6c63430008150033


Safe Address: gor:0x0a8Bb367aDa8a7A3Bd3e3EE198f4d6552a1caE41
It  is having 3 owners & threshold value as 2.

**Reply by swatijindal12** [reply]
I found that when i pass argument in constructor then it fail & produce error "GS013".
I removed constructor argument & tried then it deployed contract succesfully. 
So, is it right? that we are not able to deploy contract when constructor contain arguments?

And one more thing, when i deployed using interface then owner is createCall contctra not my gnosis safe wallet.

**Reply by DaniSomoza** [reply]
Hi @swatijindal12 

> we are not able to deploy contract when constructor contain arguments?

Yes, you can execute a constructor that takes arguments. What you need to do is append the ABI-encoded arguments to the end of the contract's bytecode when you deploy it. This combined data (bytecode + encoded arguments) becomes the initialization data. 

Once the contract is deployed, the constructor will run with the provided arguments.

Just make sure you correctly ABI-encode the constructor arguments and append them to the contract bytecode when you're preparing the initialization data.

As an example, I deployed [this contract in Goerli](https://goerli.etherscan.io/address/0x68178e9d9a99D06Dec7647716AB84126c3476791#code) that has a constructor that receives 2 arguments, an address and a uint.

You can see these 2 arguments appended at the end of the contract's bytecode:

- first param (address): `0x680cde08860141F9D223cE4E620B10Cd6741037E`
- second param (uint): `111222333` (`0x6a11e3d` in hex)

```
608060405 [...] 2064736f6c63430008000033000000000000000000000000680cde08860141f9d223ce4e620b10cd6741037e0000000000000000000000000000000000000000000000000000000006a11e3d
```


> when i deployed using interface then owner is createCall contract not my gnosis safe wallet

As I mentioned in my first comment, if you want the owner to be your Safe account, you need to use a **delegateCall** instead of a regular **call** to the createCall contract.

[In this other comment](https://github.com/safe-global/safe-apps-sdk/discussions/523#discussioncomment-7178390) I provided an example of how to achieve this using our `protocol-kit`. Set the `operation` to `OperationType.DelegateCall`:

```javascript
  const transaction = {
    to: createCallContractAddress,
    value,
    data,
    operation: OperationType.DelegateCall
  }
```

**Reply by swatijindal12** [reply]
Thank you @DaniSomoza. I tried it was working.
